src/node/app/proxy.ts

```diff
@@ -0,0 +1,190 @@
+import { logger } from "@coder/logger"
+import * as http from "http"
+import proxy from "http-proxy"
+import * as net from "net"
+import * as querystring from "querystring"
+import { HttpCode, HttpError } from "../../common/http"
+import { HttpProvider, HttpProviderOptions, HttpProxyProvider, HttpResponse, Route } from "../http"
+
+interface Request extends http.IncomingMessage {
+  base?: string
+}
+
+/**
+ * Proxy HTTP provider.
+ */
+export class ProxyHttpProvider extends HttpProvider implements HttpProxyProvider {
+  /**
+   * Proxy domains are stored here without the leading `*.`
+   */
+  public readonly proxyDomains: string[]
+  private readonly proxy = proxy.createProxyServer({})
+
+  /**
+   * Domains can be provided in the form `coder.com` or `*.coder.com`. Either
+   * way, `<number>.coder.com` will be proxied to `number`.
+   */
+  public constructor(options: HttpProviderOptions, proxyDomains: string[] = []) {
+    super(options)
+    this.proxyDomains = proxyDomains.map((d) => d.replace(/^\*\./, "")).filter((d, i, arr) => arr.indexOf(d) === i)
```

Comment:
What is the filter and indexOf doing here?

Reply:
That's filtering out items that were already seen in the array (making it unique).

Reply:
Let's use `[...new Set(array)]` instead.

Reply:
I like that; I might just keep it a set since I suppose there's no need for it to be an array.